File: python/radar_chart.py
"""P3-style Lateral Performance Indicators radar chart."""

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from math import pi


def plot_radar(athlete_row: pd.Series, title: str = "LATERAL SKATER") -> None:
    categories = [
        "L - Lateral Force",
        "R - Hip\nAbduction",
        "R - Hip Ext\nVelocity",
        "R - Lateral Force",
        "L - Hip Ext\nVelocity",
        "L - Hip\nAbduction",
    ]
    values = [
        athlete_row["l_lateral_force"],
        athlete_row["r_hip_abduction"],
        athlete_row["r_hip_ext_velocity"],
        athlete_row["r_lateral_force"],
        athlete_row["l_hip_ext_velocity"],
        athlete_row["l_hip_abduction"],
    ]

    N = len(categories)
    angles = [n / float(N) * 2 * pi for n in range(N)]
    angles += angles[:1]
    values_plot = values + values[:1]

    fig, ax = plt.subplots(figsize=(7, 7), subplot_kw=dict(polar=True))
    fig.patch.set_facecolor("#0d0d0d")
    ax.set_facecolor("#0d0d0d")

    ax.set_ylim(0, 100)
    ax.set_yticks([20, 40, 60, 80, 100])
    ax.set_yticklabels([])
    ax.yaxis.grid(color="#333333", linewidth=0.8)
    ax.xaxis.grid(color="#333333", linewidth=0.8)

    ax.plot(angles, values_plot, color="#4da6ff", linewidth=2.5)
    ax.fill(angles, values_plot, color="#4da6ff", alpha=0.15)

    ax.set_xticks(angles[:-1])
    ax.set_xticklabels(categories, color="white", fontsize=10, fontweight="bold")
    ax.tick_params(axis="x", pad=18)

    fig.text(0.5, 0.97, title, ha="center", va="top",
             color="white", fontsize=14, fontweight="bold")
    fig.text(0.5, 0.93, "Percentile", ha="center", va="top",
             color="#aaaaaa", fontsize=10)

    plt.tight_layout()
    out = f"radar_{athlete_row['athlete'].replace(' ', '_')}.png"
    plt.savefig(out, dpi=150, bbox_inches="tight", facecolor=fig.get_facecolor())
    print(f"Saved {out}")
    plt.close()


if __name__ == "__main__":
    df = pd.read_csv("../data/sample_athletes.csv")
    plot_radar(df.iloc[0])
    # To plot all athletes:
    # for _, row in df.iterrows():
    #     plot_radar(row)

File: R/radar_chart.R
# P3-style Lateral Performance Indicators radar chart
# install.packages("fmsb")
library(fmsb)

plot_radar <- function(athlete_name, data_path = "../data/sample_athletes.csv") {
  df  <- read.csv(data_path)
  row <- df[df$athlete == athlete_name, ]
  if (nrow(row) == 0) stop(paste("Athlete not found:", athlete_name))

  metrics <- data.frame(
    `L - Lateral Force`    = row$l_lateral_force,
    `R - Hip Abduction`    = row$r_hip_abduction,
    `R - Hip Ext Velocity` = row$r_hip_ext_velocity,
    `R - Lateral Force`    = row$r_lateral_force,
    `L - Hip Ext Velocity` = row$l_hip_ext_velocity,
    `L - Hip Abduction`    = row$l_hip_abduction,
    check.names = FALSE
  )

  # fmsb requires max/min rows prepended
  plot_data <- rbind(rep(100, 6), rep(0, 6), metrics)

  out_file <- paste0("radar_", gsub(" ", "_", athlete_name), ".png")
  png(out_file, width = 700, height = 700, bg = "#0d0d0d")
  par(bg = "#0d0d0d", mar = c(2, 2, 4, 2))

  radarchart(
    plot_data,
    axistype   = 1,
    pcol       = "#4da6ff",
    pfcol      = adjustcolor("#4da6ff", alpha.f = 0.15),
    plwd       = 2.5,
    cglcol     = "#333333",
    cglty      = 1,
    cglwd      = 0.8,
    axislabcol = "#555555",
    vlcex      = 0.85,
    calcex     = 0.7
  )

  title(main = "LATERAL SKATER", col.main = "white", cex.main = 1.4, font.main = 2)
  mtext("Percentile", col = "#aaaaaa", cex = 0.9)
  dev.off()
  message("Saved ", out_file)
}

plot_radar("Scottie Barnes")
